src/firmware_new/boards/frankie_v1/io/proto_multicast_communication.c

```diff
@@ -0,0 +1,230 @@
+static osEventFlagsId_t networking_event;
+
+// flag mask for network interface cofigured event
+static uint32_t NETIF_CONFIGURED = 1 << 0;
+
+// the timeout to recv a network packet
+static uint32_t NETWORK_TIMEOUT_MS = 1000;
+
+/**
+ * ProtoMulticastCommunicationProfile_t contains the information required
+ * to join a multicast group and then send or receive protobuf.
+ */
+typedef struct ProtoMulticastCommunicationProfile
+{
+    // name of the profile to pass down to other
+    // elements that may require a name (mutex, etc..)
+    const char* profile_name;
+
+    // multicast group info: the address and the port to bind to and
+    // join to send/recv proto
+    ip_addr_t multicast_address;
+    uint16_t port;
+
+    // protobuf info: the pb_field_t array fully defines the protobuf_struct,
+    // so we use a void pointer to be generic to any protobuf message.
+    void* protobuf_struct;
+    const pb_field_t* message_fields;
+    uint16_t message_max_size;
+
+    // communication_event: these events will be used to control when the networking
+    // tasks run. The networking tasks will also signal certain events.
+    osEventFlagsId_t communication_event;
+
+    // mutex to protect the protobuf struct
+    osMutexId_t profile_mutex;
+
+} ProtoMulticastCommunicationProfile_t;
+
+void io_proto_multicast_communication_init(void)
+{
+    networking_event = osEventFlagsNew(NULL);
+}
+
+ProtoMulticastCommunicationProfile_t* io_proto_multicast_communication_profile_create(
+    const char* profile_name, const char* multicast_address, uint16_t port,
+    void* protobuf_struct, const pb_field_t* message_fields, uint16_t message_max_size)
+{
+    ProtoMulticastCommunicationProfile_t* profile =
+        (ProtoMulticastCommunicationProfile_t*)malloc(
+            sizeof(ProtoMulticastCommunicationProfile_t));
+
+    const osMutexAttr_t mutex_attr = {profile_name, osMutexPrioInherit, NULL, 0U};
+
+    profile->profile_name        = profile_name;
+    profile->port                = port;
+    profile->message_fields      = message_fields;
+    profile->message_max_size    = message_max_size;
+    profile->protobuf_struct     = protobuf_struct;
+    profile->profile_mutex       = osMutexNew(&mutex_attr);
+    profile->communication_event = osEventFlagsNew(NULL);
+    ip6addr_aton(multicast_address, &profile->multicast_address);
+
+    return profile;
+}
+
+void io_proto_multicast_sender_Task(void* arg)
+{
+    ProtoMulticastCommunicationProfile_t* comm_profile =
+        (ProtoMulticastCommunicationProfile_t*)arg;
+
+    osEventFlagsWait(networking_event, NETIF_CONFIGURED, osFlagsWaitAny, osWaitForever);
+
+    // Bind the socket to the multicast address and port we then use that
+    // communication profile to join the specified multicast group.
+    //
+    // We use IP6_ADDR_ANY which will default to the ethernet interface on the STM32H7
+    struct netconn* conn = netconn_new(NETCONN_UDP_IPV6);
```

Comment:
I don't think we do, because if this task is exiting the board is either being reset or we lose power

but I just added it after the infinite for loop to be safe

Reply:
Yea lets be explicit even if it's technically not needed. I'd really prefer not to leak memory if if we think it wouldn't matter in any IRL case.

Reply:
added